sturdy/validator/reward.py

```diff
@@ -101,15 +101,25 @@ def normalize_squared(
 
 
 def calculate_penalties(
-    similarity_matrix: dict[str, dict[str, float]],
+    allocation_similarity_matrix: dict[str, dict[str, float]],
+    apy_similarity_matrix: dict[str, dict[str, float]],
     axon_times: dict[str, float],
-    similarity_threshold: float = SIMILARITY_THRESHOLD,
+    allocation_similarity_threshold: float = ALLOCATION_SIMILARITY_THRESHOLD,
+    apy_similarity_threshold: float = APY_SIMILARITY_THRESHOLD,
 ) -> dict[str, int]:
-    penalties = {miner: 0 for miner in similarity_matrix}
-
-    for miner_a, similarities in similarity_matrix.items():
-        for miner_b, similarity in similarities.items():
-            if similarity <= similarity_threshold and axon_times[miner_a] <= axon_times[miner_b]:
+    penalties = {miner: 0 for miner in allocation_similarity_matrix}
+
+    for miner_a in allocation_similarity_matrix:
+        allocation_similarities = allocation_similarity_matrix[miner_a]
+        apy_similarities = apy_similarity_matrix[miner_a]
+        for miner_b in allocation_similarities:
+            allocation_similarity = allocation_similarities[miner_b]
+            apy_similarity = apy_similarities[miner_b]
+            if (
+                allocation_similarity <= allocation_similarity_threshold
+                and apy_similarity <= apy_similarity_threshold
+                and axon_times[miner_a] <= axon_times[miner_b]
+            ):
                 penalties[miner_b] += 1
```

Comment:
HHmmmm. I guess considering the apy similarities does help with unflattening the curve further.